### PR TITLE
Implement snapshot taker

### PR DESCRIPTION
**Pull request by bgpat**
## WHY

To take a snapshot for Elasticsearch indices and restore.

## WHAT

I implemented with golang.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  - secure: "Gkd0wSMsujsDB7iTeM0VLFTGJSHRtdjvLhchlIrOJITQlI2VDBSEPvBGWSnUiQiN2zaK/xsJwM7XKG4W9ie2CFu1r4rFIvDIr1z0DQOCQghFKACTqgVz00ydg0ub3n2pxP9cygfJDb7SBGzjymgDLo+3RIGy06mJ8iIQKrl5qjiRecJS3ejTOU5wciQiTsjWzPM5u0En7QwHaKlLtoMT70ICodNIxxCa2W8HXuTMTUzeTcdBpwFTYqJWPRkqVLfmChKRwizkWnAqHcp/19LPVrWqfpExfZYJCA+zxOu+93/BaShY2czq+prYx/Vz3VjwEdWuyrZeK92/J9LBdsVGzrKJn2Qss/n5mfKr5dIs3GN3BIIgrPL+lp3/jwjDWa4cSrBa3t4rOckghd/GmscWO2W5ATGe4FQkriaHnenou7NxzmS6c1wSjEEn+2/4++eWMEEsptmUewMKlAaL9ASoDYl+AyTGuAlK5ai/05Blk985mSkgb0OYx/iD1N0ssAu2tVr7sPhxBm7q8OHMQg0YOQzMPlkgxba10c8eJT8ufslsU60ui+uJeHIkvl1BX5FZVwp2LVMATGXk+QwQ4A6npqm7dJKum4vz7MvrAZY7M/rYHkwRcvrKENnROEDu6/Ias6pxYYI0L0AbFwVW1mA0t2QsWb8fmF6MrBztcKxcl2k="
```
[the remark this answers is not in the document]
Comment (bgpat) [reply .travis.yml]: added.

**Comment by potsbo** [comment]
Would you test this with actual Elasticsearch?
I wanna see how this works and with example commands.

**Comment by bgpat** [comment]
### backup
```console
$ GOREQUEST_DEBUG=1 go run *.go -url=$ELASTICSEARCH_URL -service-name=$SERVICE_NAME -bucket=$S3_BUCKET -env=production
[http] 2018/06/20 14:07:53 HTTP Request: PUT /_snapshot/201806 HTTP/1.1
Host: localhost
Content-Type: application/json

{"settings":{"access_key":"********","base_path":"********/production/201806","bucket":"********","compress":false,"region":"********","secret_key":"********"},"type":"s3"}
[http] 2018/06/20 14:07:55 HTTP Response: HTTP/2.0 200 OK
Content-Length: 21
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: Authorization, X-Requested-With, Content-Type, Content-Length
Access-Control-Allow-Methods: OPTIONS, HEAD, GET, POST, PUT, DELETE
Content-Type: application/json; charset=UTF-8
Date: Wed, 20 Jun 2018 05:07:55 GMT
Server: nginx

{"acknowledged":true}
[http] 2018/06/20 14:07:55 HTTP Request: PUT /_snapshot/201806/20 HTTP/1.1
Host: localhost
Content-Type: application/json

{"ignore_unavailable":false,"include_global_state":false,"indices":"*"}
[http] 2018/06/20 14:07:55 HTTP Response: HTTP/2.0 200 OK
Content-Length: 17
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: Authorization, X-Requested-With, Content-Type, Content-Length
Access-Control-Allow-Methods: OPTIONS, HEAD, GET, POST, PUT, DELETE
Content-Type: application/json; charset=UTF-8
Date: Wed, 20 Jun 2018 05:07:55 GMT
Server: nginx

{"accepted":true}
[http] 2018/06/20 14:07:55 HTTP Request: GET /_snapshot/201806/20 HTTP/1.1
Host: localhost
Content-Type: application/json

[http] 2018/06/20 14:07:55 HTTP Response: HTTP/2.0 200 OK
Content-Length: 309
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: Authorization, X-Requested-With, Content-Type, Content-Length
Access-Control-Allow-Methods: OPTIONS, HEAD, GET, POST, PUT, DELETE
Content-Type: application/json; charset=UTF-8
Date: Wed, 20 Jun 2018 05:07:55 GMT
Server: nginx

{"snapshots":[{"snapshot":"20","uuid":"StZm8IOGS3662lvYWpstew","version_id":5020199,"version":"5.2.1","indices":["foo","bar"],"state":"IN_PROGRESS","start_time":"2018-06-20T05:07:55.255Z","start_time_in_millis":1529471275255,"failures":[],"shards":{"total":0,"failed":0,"successful":0}}]}
2018/06/20 14:07:55 GET https://localhost/_snapshot/201806/20:
	status_code = 200
	body = {"snapshots":[{"snapshot":"20","uuid":"StZm8IOGS3662lvYWpstew","version_id":5020199,"version":"5.2.1","indices":["foo","bar"],"state":"IN_PROGRESS","start_time":"2018-06-20T05:07:55.255Z","start_time_in_millis":1529471275255,"failures":[],"shards":{"total":0,"failed":0,"successful":0}}]}
	state = IN_PROGRESS
	shards = main.SnapshotShardInfo{Total:0, Failed:0, Successful:0}
2018/06/20 14:07:55 complete
```

### restore

```console
$ GOREQUEST_DEBUG=1 go run *.go -url=$ELASTICSEARCH_URL  -service-name=$SERVICE_NAME -bucket=$S3_BUCKET -env=production -restore
[http] 2018/06/20 14:13:08 HTTP Request: PUT /_snapshot/201806 HTTP/1.1
Host: localhost
Content-Type: application/json

{"settings":{"access_key":"********","base_path":"********/production/201806","bucket":"********","compress":false,"region":"********","secret_key":"********"},"type":"s3"}
[http] 2018/06/20 14:13:09 HTTP Response: HTTP/1.1 200 OK
Connection: close
Content-Length: 21
Content-Type: application/json; charset=UTF-8
Date: Wed, 20 Jun 2018 05:13:09 GMT
Server: nginx/1.11.9

{"acknowledged":true}
[http] 2018/06/20 14:13:09 HTTP Request: DELETE /* HTTP/1.1
Host: localhost
Content-Type: application/json

[http] 2018/06/20 14:13:09 HTTP Response: HTTP/1.1 200 OK
Connection: close
Content-Length: 21
Content-Type: application/json; charset=UTF-8
Date: Wed, 20 Jun 2018 05:13:09 GMT
Server: nginx/1.11.9

{"acknowledged":true}
[http] 2018/06/20 14:13:09 HTTP Request: POST /_snapshot/201806/20/_restore HTTP/1.1
Host: localhost
Content-Type: application/json

{"ignore_unavailable":false,"include_global_state":false,"indices":"*"}
[http] 2018/06/20 14:13:10 HTTP Response: HTTP/1.1 200 OK
Connection: close
Content-Length: 17
Content-Type: application/json; charset=UTF-8
Date: Wed, 20 Jun 2018 05:13:10 GMT
Server: nginx/1.11.9

{"accepted":true}
2018/06/20 14:13:10 complete
```

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  # GITHUB_USERNAME
+  - secure: "Gkd0wSMsujsDB7iTeM0VLFTGJSHRtdjvLhchlIrOJITQlI2VDBSEPvBGWSnUiQiN2zaK/xsJwM7XKG4W9ie2CFu1r4rFIvDIr1z0DQOCQghFKACTqgVz00ydg0ub3n2pxP9cygfJDb7SBGzjymgDLo+3RIGy06mJ8iIQKrl5qjiRecJS3ejTOU5wciQiTsjWzPM5u0En7QwHaKlLtoMT70ICodNIxxCa2W8HXuTMTUzeTcdBpwFTYqJWPRkqVLfmChKRwizkWnAqHcp/19LPVrWqfpExfZYJCA+zxOu+93/BaShY2czq+prYx/Vz3VjwEdWuyrZeK92/J9LBdsVGzrKJn2Qss/n5mfKr5dIs3GN3BIIgrPL+lp3/jwjDWa4cSrBa3t4rOckghd/GmscWO2W5ATGe4FQkriaHnenou7NxzmS6c1wSjEEn+2/4++eWMEEsptmUewMKlAaL9ASoDYl+AyTGuAlK5ai/05Blk985mSkgb0OYx/iD1N0ssAu2tVr7sPhxBm7q8OHMQg0YOQzMPlkgxba10c8eJT8ufslsU60ui+uJeHIkvl1BX5FZVwp2LVMATGXk+QwQ4A6npqm7dJKum4vz7MvrAZY7M/rYHkwRcvrKENnROEDu6/Ias6pxYYI0L0AbFwVW1mA0t2QsWb8fmF6MrBztcKxcl2k="
```
Comment (potsbo) [inline .travis.yml]: Note: it's ok to publish this.
> Encrypted variables are not available to untrusted builds such as pull requests coming from another repository.

see https://docs.travis-ci.com/user/environment-variables/#Defining-encrypted-variables-in-.travis.yml

And we have confirmed this behavior with #2

**Comment by bgpat** [comment]
thx!!!!!

**Pull request merged**